alica_engine/src/engine/RunnableObject.cpp

```diff
@@ -34,70 +38,125 @@ ThreadSafePlanInterface RunnableObject::getPlanContext() const
     return ThreadSafePlanInterface(isExecutingInContext() ? _execContext.load() : nullptr);
 }
 
-void RunnableObject::stop()
+void RunnableObject::doStop()
 {
     if (!isActive(_signalState.load())) {
         return;
     }
     ++_signalState;
-    _engine->editScheduler().schedule(std::bind(&RunnableObject::doTerminate, this));
 }
 
-void RunnableObject::start(RunningPlan* rp)
+void RunnableObject::doStart(RunningPlan* rp)
 {
     if (isActive(_signalState.load())) {
         return;
     }
     ++_signalState;
     _signalContext.store(rp);
-    _engine->editScheduler().schedule(std::bind(&RunnableObject::doInit, this));
 }
 
-void RunnableObject::setTerminatedState()
+void RunnableObject::startTrace()
 {
-    ++_execState;
-
-    clearFlags(Flags::RUN_TRACED);
-    if (!areFlagsSet(Flags::INIT_EXECUTED)) {
-        _execContext.store(nullptr);
+    if (!_engine->getTraceFactory()) {
         return;
     }
-    clearFlags(Flags::INIT_EXECUTED);
-}
 
-void RunnableObject::traceTermination()
-{
-    if (_trace) {
-        _trace->setLog({"Terminate", "true"});
-        _trace.reset();
+    switch (_tracingType) {
+    case TracingType::DEFAULT: {
+        auto parent = _execContext.load()->getParent();
+        for (; parent && (!parent->getBasicPlan() || !parent->getBasicPlan()->getTraceContext().has_value()); parent = parent->getParent());
+        _trace = _engine->getTraceFactory()->create(_name, parent ? parent->getBasicPlan()->getTraceContext() : std::nullopt);
+        break;
+    }
+    case TracingType::SKIP: {
+        break;
+    }
+    case TracingType::ROOT: {
+        _trace = _engine->getTraceFactory()->create(_name);
+        break;
+    }
+    case TracingType::CUSTOM: {
+        if (!_customTraceContextGetter) {
+            ALICA_ERROR_MSG("[RunnableObject] Custom tracing type specified, but no getter for the trace context is provided. Switching to default tracing type instead");
+            _tracingType = TracingType::DEFAULT;
+            startTrace();
+            break;
```

Comment:
nit: safer to return imo in case logic gets added after the switch statement, or just check this case when setting trace type to begin with

Reply:
I'll do the check while setting the tracing type.